#!/usr/bin/env bash

#read in variables
source CV_foureps.sh

#set lengths and offsets. These are the final ones which you will use for any DE analysis.
#Use the plots generated by the periodicity and offset scripts to decide what values to set these at

lengths='28,29,30,31,32,33'
offsets='12,12,12,13,13,13'

#run counting script to generate <.counts> files


for filename in $RPF_filenames
do
counting_script.py -bam $BAM_dir/${filename}_pc_sorted.bam -fasta $most_abundant_fasta -len $lengths -offset $offsets -out_file ${filename}_pc_final.counts -out_dir $counts_dir &
done
wait

#RPFs_9 scripts to get out counts from all regions
#run summing_CDS_counts.py script to count all reads for each CDS for use with DEseq2
#the -remove_end_codons doesn't count any reads which correspond to A-site occupation within the first 20 or last 10 codons
#use all alignments so that the most abundant transcript per gene (based on paired total RNA-seq data) can be selected downstream

for filename in $RPF_filenames
do
summing_CDS_counts.py ${filename}_pc_final.counts $region_lengths -remove_end_codons -in_dir $counts_dir -out_dir $cds_counts_dir &
done
wait

# 5UTRS

for filename in $RPF_filenames
do
summing_UTR5_counts.py ${filename}_pc_final.counts $region_lengths -in_dir $counts_dir -out_dir $UTR5_counts_dir &
done
wait

#counts to csv

for filename in $RPF_filenames
do
counts_to_csv.py ${filename}_pc_final.counts $most_abundant_fasta -all_transcripts -in_dir $counts_dir -out_dir $csv_counts_dir &
done
wait
